File: smartroom-api/app/api/v1/buildings.py
"""Bâtiments, étages et plans d'implantation."""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Body, Depends, status

from app.api.deps import (
    ROOMS_MANAGE,
    CurrentPrincipal,
    SessionDep,
    require_permission,
)
from app.api.v1.schemas import (
    BuildingIn,
    BuildingOut,
    BuildingPatchIn,
    FloorCreateIn,
    FloorOut,
    FloorPatchIn,
    FloorPlanOut,
    PlacementIn,
    RoomPlacementOut,
    UploadIn,
    VisuelIn,
)
from app.api.v1.serializers import batiment_sortie, etage_sortie
from app.core import storage
from app.services import parc_service as service

router = APIRouter(tags=["parc"])

Ecriture = Depends(require_permission(ROOMS_MANAGE))


@router.get(
    "/buildings",
    response_model=list[BuildingOut],
    summary="Lister les bâtiments",
    description=(
        "Chaque bâtiment porte ses décomptes d'étages et de salles, agrégés en "
        "SQL. La collection est courte et bornée par le parc : elle n'est pas "
        "paginée."
    ),
)
def list_buildings(session: SessionDep, _: CurrentPrincipal) -> list[BuildingOut]:
    return [batiment_sortie(*ligne) for ligne in service.list_buildings(session)]


@router.get("/buildings/{building_id}", response_model=BuildingOut, summary="Fiche bâtiment")
def get_building(
    building_id: uuid.UUID, session: SessionDep, _: CurrentPrincipal
) -> BuildingOut:
    return batiment_sortie(service.get_building(session, building_id))


@router.post(
    "/buildings",
    response_model=BuildingOut,
    status_code=status.HTTP_201_CREATED,
    summary="Déclarer un bâtiment",
    description=(
        "Le code est unique et sert d'identifiant lisible dans les exports. "
        "Un bâtiment naît sans étage : les niveaux s'ajoutent ensuite, et une "
        "salle ne se rattache qu'à un étage."
    ),
    responses={422: {"description": "Code déjà pris."}},
)
def create_building(payload: BuildingIn, session: SessionDep, _admin=Ecriture) -> BuildingOut:
    batiment = service.create_building(session, payload)
    session.commit()
    return batiment_sortie(batiment)


@router.patch(
    "/buildings/{building_id}",
    response_model=BuildingOut,
    summary="Modifier un bâtiment",
    description=(
        "Le code n'est pas modifiable : il est cité dans les exports déjà "
        "produits et dans le journal d'audit, et le changer réécrirait le passé."
    ),
)
def update_building(
    building_id: uuid.UUID, payload: BuildingPatchIn, session: SessionDep, _admin=Ecriture
) -> BuildingOut:
    batiment = service.update_building(session, building_id, payload)
    session.commit()
    return batiment_sortie(batiment)


@router.delete(
    "/buildings/{building_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Supprimer un bâtiment vide",
    description=(
        "Refusé tant que le bâtiment porte une salle. Archiver les salles en "
        "cascade serait pire qu'un refus : une salle archivée reste citée dans "
        "les réservations passées, et son bâtiment doit rester lisible."
    ),
    responses={422: {"description": "Le bâtiment porte encore des salles."}},
)
def delete_building(building_id: uuid.UUID, session: SessionDep, _admin=Ecriture) -> None:
    service.delete_building(session, building_id)
    session.commit()


@router.put(
    "/buildings/{building_id}/image",
    response_model=BuildingOut,
    summary="Déposer la photographie d'un bâtiment",
    description=(
        "PNG, JPEG ou WebP, 5 Mo au maximum. Ni PDF ni SVG : l'image s'affiche "
        "dans une carte, et le SVG porte du script qui s'exécuterait avec les "
        "droits de l'application. Le visuel précédent est effacé du disque."
    ),
    responses={422: {"description": "Format refusé, fichier vide ou trop lourd."}},
)
def upload_building_image(
    building_id: uuid.UUID, payload: VisuelIn, session: SessionDep, _admin=Ecriture
) -> BuildingOut:
    batiment = service.set_building_image(
        session, building_id, contenu=payload.content, content_type=payload.content_type
    )
    session.commit()
    return batiment_sortie(batiment)


@router.delete(
    "/buildings/{building_id}/image",
    response_model=BuildingOut,
    summary="Retirer la photographie d'un bâtiment",
)
def delete_building_image(
    building_id: uuid.UUID, session: SessionDep, _admin=Ecriture
) -> BuildingOut:
    batiment = service.delete_building_image(session, building_id)
    session.commit()
    return batiment_sortie(batiment)


@router.post(
    "/buildings/{building_id}/floors",
    response_model=FloorOut,
    status_code=status.HTTP_201_CREATED,
    summary="Ajouter un étage",
    description=(
        "`level` est un entier de tri, distinct de `code` : « RDC », « 1er » et "
        "« 2e » ne s'ordonnent pas comme du texte, et une liste triée "
        "alphabétiquement placerait le rez-de-chaussée entre le premier et le "
        "deuxième."
    ),
    responses={422: {"description": "Code ou niveau déjà pris dans ce bâtiment."}},
)
def create_floor(
    building_id: uuid.UUID, payload: FloorCreateIn, session: SessionDep, _admin=Ecriture
) -> FloorOut:
    etage = service.create_floor(session, building_id, payload)
    session.commit()
    return etage_sortie(etage)


@router.patch("/floors/{floor_id}", response_model=FloorOut, summary="Modifier un étage")
def update_floor(
    floor_id: uuid.UUID, payload: FloorPatchIn, session: SessionDep, _admin=Ecriture
) -> FloorOut:
    etage = service.update_floor(session, floor_id, payload)
    session.commit()
    return etage_sortie(etage)


@router.delete(
    "/floors/{floor_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Supprimer un étage vide",
    description="Refusé tant que l'étage porte une salle. Son plan est effacé avec lui.",
    responses={422: {"description": "L'étage porte encore des salles."}},
)
def delete_floor(floor_id: uuid.UUID, session: SessionDep, _admin=Ecriture) -> None:
    service.delete_floor(session, floor_id)
    session.commit()


@router.get(
    "/buildings/{building_id}/floors",
    response_model=list[FloorOut],
    summary="Étages d'un bâtiment",
    description="Ordonnés du sous-sol au dernier niveau.",
)
def list_floors(
    building_id: uuid.UUID, session: SessionDep, _: CurrentPrincipal
) -> list[FloorOut]:
    return [etage_sortie(*ligne) for ligne in service.list_floors(session, building_id)]


@router.get(
    "/floors/{floor_id}/plan",
    response_model=FloorPlanOut,
    summary="Plan d'un étage",
    responses={404: {"description": "Aucun plan téléversé pour cet étage."}},
)
def get_plan(floor_id: uuid.UUID, session: SessionDep, _: CurrentPrincipal) -> FloorPlanOut:
    return FloorPlanOut.model_validate(service.get_floor_plan(session, floor_id))


@router.put(
    "/floors/{floor_id}/plan",
    response_model=FloorPlanOut,
    summary="Déposer le plan d'un étage",
    description=(
        "Remplace le plan existant. Image ou PDF, 5 Mo au maximum : le fichier "
        "est servi par l'application, et accepter un type arbitraire "
        "reviendrait à héberger n'importe quel exécutable sur son domaine. Le "
        "plan précédent est effacé du disque. Le contenu voyage encodé en "
        "base64 dans le corps JSON, le multipart demandant une dépendance de "
        "plus."
    ),
    responses={422: {"description": "Format refusé, fichier vide ou trop lourd."}},
)
def upload_plan(
    floor_id: uuid.UUID, payload: UploadIn, session: SessionDep, _admin=Ecriture
) -> FloorPlanOut:
    plan = service.replace_floor_plan(
        session,
        floor_id,
        contenu=payload.content,
        content_type=payload.content_type,
        file_name=payload.file_name,
        admin_id=_admin.user_id,
    )
    session.commit()
    return FloorPlanOut.model_validate(plan)


@router.delete(
    "/floors/{floor_id}/plan",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Retirer le plan d'un étage",
)
def delete_plan(floor_id: uuid.UUID, session: SessionDep, _admin=Ecriture) -> None:
    service.delete_floor_plan(session, floor_id)
    session.commit()


@router.patch(
    "/floors/{floor_id}/placements",
    response_model=list[RoomPlacementOut],
    summary="Positionner les salles sur le plan",
    description=(
        "Coordonnées en pourcentage de la surface : elles survivent au "
        "remplacement du plan par une image de dimensions différentes, ce que "
        "des pixels ne feraient pas. Une salle archivée est refusée."
    ),
    responses={422: {"description": "Salle hors de l'étage, ou archivée."}},
)
def set_placements(
    floor_id: uuid.UUID,
    session: SessionDep,
    placements: Annotated[list[PlacementIn], Body(max_length=200)],
    _admin=Ecriture,
) -> list[RoomPlacementOut]:
    resultat = service.set_placements(session, floor_id, placements)
    session.commit()
    return [RoomPlacementOut.model_validate(item) for item in resultat]


@router.post(
    "/rooms/{room_id}/unplace",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Retirer une salle du plan",
)
def unplace(room_id: uuid.UUID, session: SessionDep, _admin=Ecriture) -> None:
    service.unplace(session, room_id)
    session.commit()

File: smartroom-front/src/api/admin/buildings.js
// src/api/admin/buildings.js
// Endpoints réels :
//   GET    /api/v1/buildings                        parc, avec décomptes
//   POST   /api/v1/buildings                        déclarer un bâtiment
//   PATCH  /api/v1/buildings/{id}                   nom, adresse, ordre
//   DELETE /api/v1/buildings/{id}                   suppression, si vide
//   PUT    /api/v1/buildings/{id}/image             photographie
//   DELETE /api/v1/buildings/{id}/image             retrait
//   GET    /api/v1/buildings/{id}/floors            étages, du bas vers le haut
//   POST   /api/v1/buildings/{id}/floors            ajouter un niveau
//   PATCH  /api/v1/floors/{id}                      renommer, renuméroter
//   DELETE /api/v1/floors/{id}                      suppression, si vide
//   GET    /api/v1/rooms?floor_id=                  salles d'un niveau
//
// L'API ne savait que *lire* le parc : les bâtiments et les étages venaient du
// jeu de démonstration, donc de nulle part une fois l'application déployée. Ce
// module sert les écritures ajoutées pour y remédier.

import * as adapt from '../adapters';
import { ApiError, del, enBase64, get, items, patch, post, put } from '../client';

/**
 * Types acceptés pour la photographie d'un bâtiment.
 *
 * Ni PDF ni SVG, bien que le magasin de médias les prenne pour les plans
 * d'étage : l'image s'affiche dans une carte, et le SVG porte du script qui
 * s'exécuterait avec les droits de l'application. Le serveur applique la même
 * liste ; celle-ci répond sans aller-retour.
 */
export const TYPES_IMAGE = ['image/png', 'image/jpeg', 'image/webp'];
const TAILLE_MAX_MO = 5;

const batiment = (data) => ({
  ...adapt.building(data),
  floorCount: data.floor_count ?? 0,
  roomCount: data.room_count ?? 0,
});

/** Le parc, du premier bâtiment au dernier selon l'ordre choisi. */
export async function listManagedBuildings({ signal } = {}) {
  const data = await get('/buildings', { signal });
  return data.map(batiment);
}

export async function getManagedBuilding(id, { signal } = {}) {
  return batiment(await get(`/buildings/${id}`, { signal }));
}

/**
 * Déclare un bâtiment.
 *
 * Le code est court, en majuscules, et sert d'identifiant lisible dans les
 * exports et le journal d'audit. Il est normalisé ici plutôt que refusé : la
 * casse n'est pas une décision de l'utilisateur.
 */
export async function createBuilding({ code, name, address, sortOrder = 0 }) {
  const identifiant = code?.trim().toUpperCase();
  if (!identifiant) throw new ApiError('Le code est obligatoire.', 422, 'code_requis');
  if (!/^[A-Z0-9]{1,4}$/.test(identifiant)) {
    throw new ApiError(
      'Le code tient en quatre caractères, lettres majuscules ou chiffres.',
      422,
      'code_invalide',
    );
  }
  if (!name?.trim()) throw new ApiError('Le nom est obligatoire.', 422, 'nom_requis');

  return batiment(
    await post('/buildings', {
      code: identifiant,
      name: name.trim(),
      address: address?.trim() || null,
      sort_order: Number(sortOrder) || 0,
    }),
  );
}

/**
 * Modifie un bâtiment.
 *
 * Le code n'y figure pas : il est cité dans les exports déjà produits et dans
 * le journal d'audit, et le changer réécrirait le passé.
 */
export async function updateBuilding(id, { name, address, sortOrder }) {
  const corps = {};
  if (name !== undefined) corps.name = name.trim();
  if (address !== undefined) corps.address = address?.trim() || null;
  if (sortOrder !== undefined) corps.sort_order = Number(sortOrder) || 0;
  return batiment(await patch(`/buildings/${id}`, corps));
}

/** Suppression, refusée par le serveur tant qu'une salle y subsiste. */
export async function deleteBuilding(id) {
  await del(`/buildings/${id}`);
  return { id, deleted: true };
}

export async function uploadBuildingImage(id, file) {
  if (!file) throw new ApiError('Aucun fichier sélectionné.', 422, 'fichier_manquant');
  if (!TYPES_IMAGE.includes(file.type)) {
    throw new ApiError(
      'Format refusé : déposez une image PNG, JPEG ou WebP.',
      422,
      'format_invalide',
    );
  }
  if (file.size > TAILLE_MAX_MO * 1024 * 1024) {
    throw new ApiError(`Fichier trop lourd : ${TAILLE_MAX_MO} Mo maximum.`, 422, 'trop_lourd');
  }

  return batiment(
    await put(`/buildings/${id}/image`, {
      content_type: file.type,
      content: await enBase64(file),
    }),
  );
}

export async function removeBuildingImage(id) {
  return batiment(await del(`/buildings/${id}/image`));
}

/**
 * Étages d'un bâtiment, chacun avec les salles qu'il porte.
 *
 * Les salles accompagnent l'étage plutôt que d'être chargées au clic : la
 * question que l'écran répond est « que contient ce bâtiment », et y répondre
 * en deux temps ferait clignoter la réponse.
 */
export async function listFloorsWithRooms(buildingId, { signal } = {}) {
  const etages = await get(`/buildings/${buildingId}/floors`, { signal });

  const salles = await Promise.all(
    etages.map((etage) =>
      get('/rooms', { params: { floor_id: etage.id, size: 100 }, signal })
        .then((page) => items(page).map(adapt.room))
        .catch(() => []),
    ),
  );

  return etages.map((etage, index) => ({
    id: etage.id,
    buildingId: etage.building_id,
    code: etage.code,
    label: etage.label,
    level: etage.level,
    roomCount: etage.room_count ?? 0,
    rooms: salles[index],
  }));
}

/**
 * Ajoute un niveau.
 *
 * `level` est un entier de tri distinct du code : « RDC », « 1er » et « 2e »
 * ne s'ordonnent pas comme du texte, et une liste triée alphabétiquement
 * placerait le rez-de-chaussée entre le premier et le deuxième.
 */
export async function createFloor(buildingId, { code, label, level }) {
  if (!code?.trim()) throw new ApiError('Le code de l’étage est obligatoire.', 422, 'code_requis');
  if (!label?.trim()) throw new ApiError('Le libellé est obligatoire.', 422, 'libelle_requis');
  const niveau = Number(level);
  if (!Number.isInteger(niveau)) {
    throw new ApiError('Le niveau est un nombre entier.', 422, 'niveau_invalide');
  }

  const data = await post(`/buildings/${buildingId}/floors`, {
    code: code.trim(),
    label: label.trim(),
    level: niveau,
  });
  return { id: data.id, code: data.code, label: data.label, level: data.level, rooms: [] };
}

export async function updateFloor(floorId, { code, label, level }) {
  const corps = {};
  if (code !== undefined) corps.code = code.trim();
  if (label !== undefined) corps.label = label.trim();
  if (level !== undefined) corps.level = Number(level);
  const data = await patch(`/floors/${floorId}`, corps);
  return { id: data.id, code: data.code, label: data.label, level: data.level };
}

export async function deleteFloor(floorId) {
  await del(`/floors/${floorId}`);
  return { id: floorId, deleted: true };
}
